README.md
# mapspec #

`mapspec` stands for **M**cmc **A**lgorithm for **P**arameters of **SPEC**tra.  It was originally designed to rescale astronomical spectra to a standard flux scale, using a method similar to that of [van Groningen & Wanders 1992](http://adsabs.harvard.edu/abs/1992PASP..104..700V).  Full details of the method are in [Fausnaugh 2017](https://ui.adsabs.harvard.edu/abs/2017PASP..129b4007F/abstract), published in  PASP (please cite this paper if you use the code).  Along the way, several spectra analysis tools useful for dealing with bright/broad emission lines have been implemented.

* * *

# Installation #

`mapspec` is a fairly simple python package.  Download it with `git clone`, `cd` to the mapspec directory, and run

```
python setup.py sdist
pip install -e ./
```

This installs to your full system---consider using a virtual environment (virtualenv or conda) or 

```
python setup.py sdist
pip install ./ --user <MAPSPECDIR>
```

for more fine-grained control.  

Alternatively, put the mapspec directory in your PYTHONPATH or your current working directory.  

`mapspec` is compatible with python 2.7 and python 3.6 or later.

To test the installation, go to `examples` and try running the regression test script:

    cd ../mapspec_test
    bash run_test.sh

`run_test.sh` runs the regression test---it runs mapspec on some sample data and checks that the outputs match standardized files in `mapspec_test/output_checks`.  The tests take about 10 minutes to run, and the results can be plotted with `mapspec_test/plot_test.py`.  More details are in `mapspec_test/README`.

# Quick Start #

The workhorse script is `script/do_map`, which is available from the command line after installing `mapspec` and has a help function.

Here are the data files that you need:

* A reference spectrum in 3 column ascii format (wavelength, flux, and flux error).
* A file with wavelength regions for the emission line.  This should have 3 rows: 1)  the wavelength window of the emission line, 2) the wavelength window of the blueward continuum, and 3) the wavelength window of the redward continuum.  Each line has 2 numbers: the blue edge of the wavelength region window and the red edge of the wavelength region window.
* A list of files with spectra to rescale, 1 file per line.  Like the reference, the spectra should be in 3 column ascii format.

Examples of all of these files are in `examples/mapspec_test`.

`do_map` runs a list of spectra through the rescaling model.  This is the normal _modus operandi_ for reverberation mapping studies, so `do_map` may have everything that most users need.  The output (using the Gauss-Hermite smoothing kernel) will be saved in files called 'scale.h._<input_spectrum_name>'.

To see how `do_map` works, you can look at the source in `mapspec/scripts/do_map`.

`do_map` also does some simple model comparisons (Gaussian smoothing vs. Gauss-Hermite smoothing).   The file `mapspec.params` compares the chi^2 from two different different rescaling models, and two output spectra are saved by default ('scale_<input_spectrum_name>' for Gaussian smoothing and  'scale.h._<input_spectrum_name>' for Gauss-Hermite smoothing).


* * *

# Outline #

`mapspec` is object-oriented and is designed to be modular and highly extensible.  It makes heavy use of numpy and scipy libraries, and was designed to easily integrate with these packages.

There are two main parts of the package:  First, general-usage spectrum utilities, which are bundled in `mapspec/spectrum.py`.  Second, the rescaling procedure, which is kept in `mapspec/mapspec.py`.  

Some helpful scripts for constructing a reference spectrum are also provided----these are `make_ref` and `smooth_ref`.  Check the help menus and internal comments for details.  By default, they are installed to your system and available on the command line.

We begin by describing (with illustrative examples) the spectrum utilities.

* * *

## Spectrum ##

The primary object is the `Spectrum` class.  `Spectrum` has 3 fundamental attributes:  wavelength (`Spectrum.wv`), flux  (`Spectrum.f`), and flux error (or measurement error, `Spectrum.ef`).  Each attribute is a numpy array, so all the functionality of numpy/scipy is preserved.  For example, to get the mean flux-level:

    from mapspec.spectrum import Spectrum
    my_spec = Spectrum()
    print my_spec.f.mean()

or to get the number of pixels:

    print my_spec.wv.size

(to actually put data in the `Spectrum` object, see "Reading Data" below).  `Spectrum` also has many useful methods to interpolate, rebin, extinction-correct, and smooth (or log-smooth) the data.  For example, this will rebin the spectrum to twice the original wavelength spacing:

    import numpy as np
    dx = my_spec.wv[1] - my_spec.wv[0] #assumes even wavelength spacing throughout
    xnew = np.arange(my_spec.wv.min(), my_spec.wv.max() + 2.*dx, 2.*dx)
    my_spec.rebin(xnew)



* * *

## Emission Lines ##

The next object is the `EmissionLine` class.  `EmissionLine` inherits from `Spectrum`, so it can be used in approximately the same way.  To instantiate an `EmissionLine`, you need to specify the line window and surrounding continuum:

    from mapspec.spectrum import EmissionLine
    oiii = [ [5065, 5115],     #OIII 5007, at redshift 1.018
             [5059, 5065],     #blue continuum
             [5115, 5129]  ]   # red continuum
    my_line = EmissionLine(my_spec, oiii[0], [oiii[1],oiii[2]] )

At instantization, `EmissionLine` fits a linear model to the local continuum (`oiii[1]` and `oiii[2]`) and subtracts the model underneath the emission line.  That is, `my_line.wv` is the wavelengths of `my_spec` between `oiii[0,0]` and `oiii[0,1]` and `my_line.f` is the corresponding section of `my_spec.f` minus the continuum fit.

`EmissionLine` also has other useful methods, such as `my_line.integrate_line()`, which will use `scipy.integrate.simps` to return the total line flux, or `my_line.fwhm()`, which will return the full-width-at-half-maximum of the line.

* * *

## Line Model ##

The last class is `LineModel`.  At instantization, `LineModel` fits a function to an input `EmissionLine`:

    from mapspec.spectrum import LineModel
    my_lmodel = LineModel(my_line, func='gauss-hermite')

The _raison d'être_ of `LineModel` is to generate flux predictions at arbitrary wavelengths.  This is accomplished by overloading the class's call method, so you can think of `LineModel` as a kind of function:

    xeval = np.linspace( my_line.wv.min(), my_line.wv.max(), 500 )
    fpredict = my_lmodel(xeval)

You could use these predictions to, for example, subtract the line from the original spectrum:

    mask = ( my_spec.wv > oiii[0,0])*(my_spec.wv < oiii[0,1] )
    my_spec.f[mask] -= my_lmodel(my_spec.wv[mask])

If the parameters of the fit are of interest, you can access then with `my_lmodel.p` (and the covariance matrix, for errors, is stored in `my_lmodel.covar`), but you will have to check the individual functions/methods to see which parameter is which.

* * *
## Reading Data ##

Input from a file is handled with inheritance.  If you know what format your data are in, the following classes know how to read these formats and return fully functioning `Spectrum` objects:

    from mapspec.spectrum import TextSpec, TextSpec_2c, FitsSpec
    
    s1 = TextSpec('mockdata.txt')  #for 3 column format
    s2 = TestSpec_2c('mockdata_2col.txt') #for 2 columns, assumes no errors
    s3 = FitsSpec('mockdata.fits')

`FitsSpec` contains many keywords in order to deal with the vast diversity of .fits formats:

    s3 = FitsSpec('mockdata.fits', extension = 1, data_axis = 1, error_axis = 2,
                  x1key = 'CRVAL1',
                  dxkey = 'CD1_1',
                  p1key = 'CRPIX1'
                  )

Two files with test data from NGC 5548 (3 column ascii) are in `examples` (test.LBT.dat and test.MDM.dat---test.dat is a copy of test.LBT.dat), and 6 spectra for rescaling are in `examples/mapspec_test`

* * *

## Rescaling Spectra ##

The tools for rescaling spectra are in `mapspec/mapspec.py`.  Most of the work is done by the `RescaleModel` class:

    from mapspec.spectrum import *
    from mapspec.mapspec import RescaleModel
    
    sref = TextSpec('ref.txt')
    lref = EmissionLine(sref, oiii[0], [ oiii[1], oiii[2] ] )

    my_model = RescaleModel(lref, kernel = 'Gauss')

`my_model` will take an input emission line and try to align it with `lref`.  The model consists of a wavelength shift, rescaling factor, and smoothing---the form of the smoothing kernal is specified by the `kernel` keyword (in the above example, a simple Gaussian).

The parameters of the model are stored in a dictionary, so they are easy to access.  If you know what parameters you want, you can immediate apply the model to the data:

    #shift by 0.25 x-units, multiply by 2, and smooth with kernel of width 1 x-units
    my_model.p = {'shift': 0.25, 'scale': 2.0, 'width': 1.0} 
    sout, lost_pix = my_model.output(my_spec)

`lost_pix` is a boolean array that tells you how much of the original spectrum is lost---you loose at least one pixel on either side for any non-zero shift.  If you don't care about the original wavelength grid, it is suggested to change the wavelengths in place (i.e., `my_spec.wv -= shift`).

Usually, you will want to optimize these parameters so that they match the reference.  To do this, a function `metro_hast` is included:

    from mapspec.mapspec import metro_hast
    chi2, p_best, frac_accept = metro_hast(5000, my_line, my_model)

This will run an MCMC for 5000 steps, and try to optimize `my_model.p` so that `my_line` matches `lref` (as a reminder, `lref` was specified when instantiating `my_model`---any `RescaleModel` object should probably have its own reference line).  `chi2` is the best (minimum) chi^2, `p_best` are the corresponding parameters, and `frac_accept` is the acceptance fraction of the chain.

You then apply the model as above:

    my_model.p = p_best
    sout,lost_pix =  my_model.output(my_spec)

`metro_hast` has some other useful options:

    chi2, p_best, frac_accept, p_chain = metro_hast(5000, my_line, my_model, keep = True)

This will return the MCMC chain of parameters in `p_chain`.  `p_chain` is actually its own object (`mapspec.mapspec.Chain`), but it is little more than a glorified NxM array that knows how to plot itself (it also stores the log-likelihood for each set of parameters in a separate attribute). Also

    plt.ion()
    chi2, p_best, frac_accept, p_chain = metro_hast(5000, my_line, my_model, plot = True)

will plot the MCMC chain as it runs.

# Details #

## Warning on Interpolation ##

At the moment, only linear interpolation supports a rigorous error propagation.  This includes calculating the covariance matrix due to both interpolation and smoothing when rescaling the data.

Other interpolation methods are supported---any 'kind' keyword for `scipy.interpolate.interp1d` can be used.  There is also an implementation of bsplines using `scipy.interpolate`, and a custom built sinc interpolator (with a Lanczos window by default).

These alternative methods do not support full error propagation.  For now, they will perform the same operation as called for on the flux spectrum, but on the *square* of the error spectrum (i.e., the variance).  This usually results in errors similar to the input and roughly preserves the fractional uncertainty.  However, we note that this is not strictly correct, and may be a feature that we improve in the future.

## Calculation of the Likelihood and Fitting Procedure ##

`mapspec` aligns the data and the model by minimzing: (data - model)^2/error^2, where data is the input line, model is the reference value chosen at instantization, and error^2 = (reference error)^2 + (data error)^2.  In other words, `mapspec` takes a maximum likelihood approach, assuming normally distributed residuals and independent uncertainties on the data and reference.

Because of wavelength shifts and edge-effects from convolution, there can be problems aligning the edges of the fitting region---in order to censor these points, it would be necessary to change the amount of data, and therefore the number of degrees-of-freedom, during the fit.  To prevent this while mitigating the edge-effects, `mapspec` ignores 10% of the data (5% on each end) when calculating the likelihood.  Although this is not an optimal solution, it does produces reasonable results and appears to be a good compromise between correctness, simplicity,  and practicability.

It is therefore suggested to choose the fitting region (line wavelengths) slightly larger than would naively be expected for isolating the emission line flux.  It is also suggested to remove large shifts (greater than 1 pixel) before performing the fit.  A convenience function `mapspec.mapspec.get_cc` is provided to cross correlate two spectra---see `do_map.py` for an example.

In the Bayesian framework, priors are always added into the likelihood calculation.  `mapspec` assumes uniform priors on all parameters, so the priors do not explicitly enter into the calculation of the log-likelihood.  Note that a flat prior *is informative* for the case of the scaling parameter---however, because the rescaling is a factor of a few (not many orders of magnitude), this situation makes little difference.

When using a Gauss-Hermite Kernel, h3 and h4 are constrained to be > -0.3 and < 0.3; experimentation shows that this range is sufficient to produce a wide range of emission line shapes.  The finite interval is enforced by setting the prior probability to 0 outside of this interval.  Similarly, strange behavior will occur if the kernel width drops below the spectral resolution (a kind of undersampling/aliasing effect).  A minimum kernel width is imposed at 1/2 of the wavelength spacing of the reference data.

`RescaleModel` objects have the option to add priors, either through a function provided by the user or from posterior distributions stored in `Chain` objects.  See `do_map.py` for an example.

* * *
# Methods #

## Spectrum ##

| Object | Method | args | Description |
|--------|--------|------|-------------|
|`Spectrum`| `__init__`| `style = 'linear'`| Change the interpolation style at instantization (implicitly calls `set_interp`).|
|`Spectrum`|`restore`|    | Restore attributes (`Spectrum.wv`,`Spectrum.f`, and `Spectrum.ef`) to values when spectrum was instantiated.|
|`Spectrum`|`interp` | `xnew`| Returns interpolated flux and flux errors at location of array `xnew`.|
|`Spectrum`|`set_interp`| kwargs| Change interpolation style (see `style` below).  Keywords for custom b-splines and sinc interpolation are also provided here.|
|`Spectrum`|`rebin`| `xnew` | Rebins the spectrum to wavelengths `xnew`.  Modifies `Spectrum.wv`,`Spectrum.f`, and `Spectrum.ef`.|
|`Spectrum`|`extinction_correct`| `E_BV`, `RV`| De-extinguish the spectrum for a [CCM89](http://adsabs.harvard.edu/abs/1989ApJ...345..245C) extinction law, for a given E(B-V) (`E_BV`) and R_V (`RV`).  Modifies `Spectrum.f`, and `Spectrum.ef`.|
|`Spectrum`|`smooth`| `width`,`name`| Smooths the spectrum with a kernel of width `width` specified with `name` (calculated with `scipy.signal.get_window`).  Modifies `Spectrum.f`, and `Spectrum.ef`.  Edge-effects are treated by replacing with the original spectrum (a number of pixels equal to `width` on both edges).|
|`Spectrum`|`velocity_smooth`| `v_width`| Smooths the spectrum with a velocity dispersion of width `v_wdith` (in km/s) by rebinning evenly in log wavelength.  Modifies `Spectrum.f`, and `Spectrum.ef`.  Converts back to original wavelengths, but cannot support sinc interpolation (unevenly spaced wavelengths).|
|`Spectrum`| `wv`| Attribute| numpy array of wavelengths.|
|`Spectrum`| `f`| Attribute| numpy array of fluxes.|

CHANGELOG.md
# Changelog

Human-readable summary of major changes and features/data that are added or removed.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
### Added
### Removed


## [0.2.0] - 2022-02-09
Reviewed for issues with the code, removed dangling branches, added python help menus and dependencies in setup.py.
### Changed
- Formalized regression tests, added annotations to plots and explicit checks.  Restructed these directories.
### Added
- Added `install_requires` to `setup,py`, changed install instructions.
- Added argparse to `do_map`, `make_ref`, and `smooth_ref`, as well as help functions.  Note that this changes the API.
### Removed
- Various shells like `run_map.sh`, etc, which remove some of the indirection.



mapspec.py
import scipy as sp
from scipy.integrate import simps
from scipy import linalg
import matplotlib.gridspec as gridspec
import matplotlib.pyplot as plt
from spectrum import *
from copy import deepcopy
#these are 'probalists', turns out to matter in order to match van der
#Marel & Franx 1993
from numpy.polynomial.hermite_e import HermiteE as H
import re


__all__ = ["RescaleModel","Chain","get_cc","metro_hast"]

debug = True


class RescaleModel(object):
    """
    This model stores parameters (shift, scale, and convolution
    kernel), performs the operation, and evaluates the likelihood.

    There are two modes for fitting----either chi^2 is calculated from
    the full covariance matrix (fit_with_covar=True), or chi^2 is
    calculated from just data errors (fit_with_covar=False).

    see do_map.py for examples of how to use.
    """

    def __init__(self,Lref,kernel='Hermite',fit_with_covar=False):
        """
        Constructs a rescaling model.  Needs a reference EmissionLine
        object, to which it will try to align data, a choice of
        functional form for the smoothing kernel, and if you want to
        use the covariance matrix during the fit.
        """
        #this is an emission line object, to which we scale to 
        self.Lref = Lref  

        self.use_covar = fit_with_covar
        #empty dictionary that will hold prior distributions.  keys
        #will be same as parameters, value will be a function that
        #evaluates the prior probability at the input parameter.
        self.prior_prob = {}
        self.kernelname = kernel

        if kernel == 'Delta':
            self._get_kernel = lambda x: sp.array([1.0])
            self.p = {'shift':1.0e-4, 'scale':1.0}
            self.set_scale = {'shift':0.05, 'scale':0.02}

        elif kernel == 'Gauss':
            self._get_kernel = self._Gauss

            #params are shift, scale, and width of gaussian
            #convolution kernel
            width_min = 0.51*(self.Lref.wv[1] - self.Lref.wv[0])
            self.p         = {'shift':1.0e-4, 'scale':1.00, 'width': width_min}
            self.set_scale = {'shift':0.05, 'scale':0.02, 'width':0.30}


        if kernel == 'Hermite':
            self._get_kernel = self._Hermite

            #params are shift, scale, width, h3, and h4 of
            #gauss-hermite polynomials
            width_min = 0.46*(self.Lref.wv[1] - self.Lref.wv[0])
            self.p         = {'shift':1.e-4, 'scale':1.00, 'width': width_min,
                              'h3':0.0, 'h4':0.0}
            self.set_scale = {'shift':0.05, 'scale':0.02, 'width':0.30,
                              'h3':0.03, 'h4':0.03}

    def __call__(self,L):
        """
        Overloaded the __call__ method so that you just use the model
        like a function to get the chi^2.  Input is an EmissionLine
        object, which will be matched to the reference.
        """

        if self.use_covar:
            y,var,mask,covar  = self._get_yz(L)
            lnlikely  = self._get_chi2_covar(y,covar,mask)
        else:
            y,var,mask  = self._get_yz(L,getcovar=False)
            lnlikely  = self._get_chi2(y,var,mask)

        lnlikely += self._add_priors()

        #take care of limited parameter space by setting prior
        #probability to 0 [ -ln(prob) = inf]
        lnlikely += self._prior_limits()

        return lnlikely


    def _get_chi2(self,y,v,m):
        return sp.sum(    (self.Lref.f[m] - y)**2/(self.Lref.ef[m]**2 + v))

    def _get_chi2_covar(self,y,C,m):
        vectoruse = sp.matrix(self.Lref.f[m] - y)
        Cuse = sp.matrix(C)
        return sp.ravel(vectoruse*linalg.inv(Cuse)*vectoruse.T)[0]


    def _add_priors(self):
        prior = 0
        for key in self.prior_prob.keys():
            prior += -2.*sp.log(
                self.prior_prob[key](self.p[key]) 
                )

        return prior


    def make_dist_prior(self,C,pname,burn = 0.5):
        params = sp.transpose(C.pchain)
        prior_dist = params[  C.index[pname]   ]
        icut = prior_dist.size*burn
        prior_dist = prior_dist[icut::]
        c1,m,c2 = sp.percentile(prior_dist,[16,50,84])
        print c1,m,c2

        #will model the distribution as a  gaussian, for now....
        self.prior_prob[pname] = lambda x: sp.exp(-0.5*(x - m)**2/ ( (c2-c1)/2 )**2 )
        self.p[pname] = m
        

    def make_func_prior(self,pname,func,params):
        self.prior_prob[pname] = lambda x: func(x,params) 


    def output(self,S,getcovar=True):
        """
        This will apply the rescaling model to a full spectrum.
        Default is to output the covariance matrix as well.
        """
        s = deepcopy(S)

        #shift
        s.wv -= self.p['shift']
        m = (S.wv >= s.wv.min()  )*(S.wv <= s.wv.max()  )
        y,z = s.interp(S.wv[m])
        
        #convolve
        k = self._get_kernel(S.wv[m])
        if getcovar:
            if self.kernelname == 'Delta':
                covar = get_covarmatrix(s.wv, S.wv[m], s.ef, k, 2)
            else:
                covar = get_covarmatrix(s.wv, S.wv[m], s.ef, k, 5*self.p['width'])

        z = sp.sqrt(sp.convolve(z**2,k**2,mode='same'))
        y = sp.convolve(y,k,mode='same')

        #scale
        z *= self.p['scale']
        y *= self.p['scale']
        if getcovar:
            covar *= self.p['scale']*self.p['scale']


        #edge effects left in for now
        sout = Spectrum()
        sout.wv = deepcopy(S.wv[m])
        sout.f  = y
        sout.ef = z

        if getcovar:
            return sout,m,covar
        else:
            return sout,m

    def _get_yz(self,L,getcovar=True):
        """
        Gets flux and error (y and z) near the line wavelengths, makes
        sure everything is aligned.
        """

        l = deepcopy(L)

        #shift
        l.wv -= self.p['shift']
        m = (self.Lref.wv >= l.wv.min() )*(self.Lref.wv <= l.wv.max() )
        y,z = l.interp(self.Lref.wv[m])
        #convolve
        k = self._get_kernel(self.Lref.wv[m])

        #have to make covar before smoothing z
        if getcovar:
            if self.kernelname == 'Delta':
                covar = get_covarmatrix(l.wv, self.Lref.wv[m], l.ef, k, 2)
            else:
                covar = get_covarmatrix(l.wv, self.Lref.wv[m], l.ef, k, 5*self.p['width'])

        z = sp.sqrt(sp.convolve(z**2,k**2,mode='same'))
        y = sp.convolve(y,k,mode='same')

        #scale
        z *= self.p['scale']
        y *= self.p['scale']
        if getcovar:
            covar *= self.p['scale']*self.p['scale']
        #add reference errors now to simplify the chi2----these don't
        #seem to matter much
            covar[sp.diag_indices_from(covar)] += self.Lref.ef[m]**2

        #trim 10% of data to help with edge effects and shifting the
        #data.  This number is hard-coded so that the degrees of
        #freedom are fixed during the fit.
        trim = round(0.05* (self.Lref.wv[m].size)) 


        z = z[trim:-trim]
        y = y[trim:-trim]
        if getcovar:
            covar = covar[:,trim:-trim]
            covar = covar[trim:-trim,:]

        #need a mask for reference when calculating chi^2
        m2 = (self.Lref.wv >= self.Lref.wv[m][trim] )*(self.Lref.wv < self.Lref.wv[m][-trim] )


        if getcovar:
            # Note that z**2 (error spectrum) is slightly different
            # than the diagonal of covar, because covariance was
            # ignored for z
            return y,z**2,m2,covar
        else:
            return y,z**2,m2

    def _Gauss(self,x):
        """
        Gaussian Smoothing kernel.
        """
        dlambda = x[1] - x[0]
        pixwidth = self.p['width']/dlambda
        prange = sp.r_[ -(x.size //2) + 1 : (x.size)//2  ]
        assert prange.size %2 == 1

        k = sp.exp(-0.5* prange**2/pixwidth**2)
        k /= abs(sp.sum(k))
        return k

    def _Hermite(self,x):
        """
        Gauss-Hermite Smoothing kernel.
        """
        dlambda = x[1] - x[0]
        pixwidth = self.p['width']/dlambda
        prange = sp.r_[ -(x.size //2) + 1 : (x.size)//2  ]
        
        h = H([ 1.0, 0.0, 0.0,self.p['h3'], self.p['h4'] ])
        #although the constants will divide out when normalizing the
        #kernel, they are important for making sure that h3 and h4 are
        #defined correctly.  Technically, this only matters for
        #choosing good intervals of h3 and h4, otherwise it is just a
        #mismatch of units

        #Equations are defined in van der Marel & Franx 1993
        k = 1./pixwidth/sp.sqrt(2*sp.pi)*sp.exp(-0.5*prange**2/pixwidth**2)*h(prange/pixwidth)
        k /= abs(sp.sum(k))


        return k


    def step(self):
        pout = {}
        for key in self.p.keys():
            pout[key] = self.p[key] + self.set_scale[key]*sp.randn()

        return pout

    def _prior_limits(self):
        prior = 0
        if len(self.p) > 2:
            dlambda = self.Lref.wv[1] - self.Lref.wv[0]
            #if width is too small, the kernel is undersampled.  Weird
            #things will happen, so this represents a lower limit.
            if self.p['width']/dlambda <0.5:
                prior = sp.inf

        if len(self.p) > 3:
            #experiments have found that h3 and h4 between -0.3 and
            #0.3 should be adequate (very diverse line shapes appear)
            if self.p['h3'] < -0.3:
                prior = sp.inf
            elif self.p['h3'] > 0.3:
                prior = sp.inf
            if self.p['h4'] < -0.3:
                prior = sp.inf
            elif self.p['h4'] > 0.3:
                prior = sp.inf

        return prior




class Chain(object):
    """
    This is an object for storing MCMC chains.  It names the
    parameters by storming them in a dictionary.  It knows how to read
    and write itself, and how to plot both histograms and triangle
    (correlation) plots.
    """
    def __init__(self):
        self.pchain   = []
        self.lnlikely = []

        self.index   = {}
        self.figure = None

#        self.figure,(self.axes) = plt.subplots(len(pnames) + 1,1)

    def add(self,M,chi2):
        if len(self.pchain) == 0:
            for i,k in enumerate(M.p.keys() ):            
                self.index[k] = i

            self.figure,(self.axes) = plt.subplots( len(M.p.keys()) + 1,1)

        self.pchain.append(deepcopy( M.p.values() ))
        self.lnlikely.append(chi2)

    def save(self,ofile):
        head = 'lnlikely   '
        outindex = []
        for key in self.index.keys():
            head += key+'   '
            outindex.append(self.index[key])
        sp.savetxt(ofile,sp.c_[self.lnlikely,sp.array(self.pchain)[:,outindex]],header=head)

    def read(self,ifile):
        fin = open(ifile,'r')
        line = fin.readline()
        pname = re.split('   ',line)
        if pname[0] != '# lnlikely':
            raise ValueError('Note a mapspec chain file! (must begin with lnlikely)')

        input_chain = sp.genfromtxt(ifile)
        self.lnlikely = input_chain[:,0]
        self.pchain   = input_chain[:,1::]
        for i,p in enumerate(pname[1:-1]):
            self.index[p] = i

        assert max(self.index.values()) == sp.transpose(self.pchain).shape[0] - 1


    def burn(self,frac):
        assert frac < 1
        cuti = int(frac*self.pchain.shape[0])
        self.pchain = self.pchain[cuti::]
        


    def plot(self,interact = 1):
        plotp = sp.transpose(self.pchain)
        if self.figure is None:
            self.figure,(self.axes) = plt.subplots( sp.transpose(self.pchain).shape[0] + 1,1)


        for ax in self.axes: ax.cla()

        self.axes[0].plot(self.lnlikely)
        self.axes[0].set_ylabel('ln likelihood')

        for key in self.index.keys():
            self.axes[self.index[key] + 1].plot(
                plotp[ self.index[key] ] 
                )
            self.axes[self.index[key] + 1].set_ylabel(key)

        for ax in self.axes[0:-1]:
            ax.set_xticks([])
        self.figure.subplots_adjust(hspace=0)
        if interact == 1:  
            plt.draw()
        return

    def plot_hist(self):
        plotp = sp.transpose(self.pchain)
        if self.figure is None:
            self.figure,(self.axes) = plt.subplots( sp.transpose(self.pchain).shape[0] + 1,1)
        for ax in self.axes: ax.cla()
        
        self.axes[0].hist(self.lnlikely,bins = 0.01*len(self.lnlikely))
        self.axes[0].set_xlabel('ln likelihood')
        for key in self.index.keys():
            self.axes[self.index[key] + 1].hist(
                plotp[ self.index[key] ] , bins = 0.01*len(plotp[self.index[key]])
                )
            self.axes[self.index[key] + 1].set_xlabel(key)

#        self.figure.set_size_inches(8,5)
        self.figure.tight_layout()

    def plot_corr(self):
        plotp = sp.transpose(self.pchain)
        if self.figure is None:
            self.gs1 = gridspec.GridSpec(plotp.shape[0],plotp.shape[0])

        for k1 in self.index.keys():
            for k2 in self.index.keys():
                if self.index[k1] > self.index[k2]: continue
                ax = plt.subplot(
                    self.gs1[
                        plotp.shape[0]-self.index[k1] - 1, plotp.shape[0]- self.index[k2] -1
                        ]
                    )

                if k1 == k2:
                    ax.hist(plotp[self.index[k1] ],0.01*len(plotp[ self.index[k1] ]),facecolor='k',alpha=0.5)
                else:
                    ax.plot(plotp[self.index[k2]],plotp[self.index[k1]],'ko',ms=2,rasterized=True,alpha=0.15)

                if self.index[k2] == plotp.shape[0] - 1:
                    ax.set_ylabel(k1)
                    if self.index[k1] != 0:
                        ax.set_xticklabels([])
                        ax.tick_params(labelright=True)
                        ax.set_yticklabels([])
                    else:
                        ax.set_xlabel(k2)

                elif self.index[k1] == 0:
                    ax.set_xlabel(k2)
                    if self.index[k2] == 0:
                        ax.yaxis.tick_right()
                    else:
                        ax.set_yticklabels([])


                elif self.index[k1] == self.index[k2]:
                    ax.yaxis.tick_right()
                    if self.index[k1] != plotp.shape[0] - 1:
                        ax.set_xticklabels([])



                else:
                    ax.set_xticklabels([])
                    ax.set_yticklabels([])



        plt.gcf().subplots_adjust(hspace=0,wspace=0)




def get_cc(y1,y2,x1,x2):
    """
    Easy way of estimating the shift to the nearest pixel (given in
    wavelength units).

    y1 = flux of spec1
    y2 = flux of spec2
    x1 = wavelengths of spec1
    x2 = wavelengths of spec2
    """

    if x1.size > x2.size:
        m = (x1 > x2.min())*(x1 < x2.max())
        yuse = [y1[m],y2]
    else:
        m = (x2 > x1.min())*(x2 < x1.max())
        yuse = [y1,y2[m]]


    cc = sp.correlate(yuse[0],yuse[1],mode='same')
    i  = sp.where(cc == cc.max())[0]
    shift = (x1[1] - x1[0] )*(cc.size//2 - i)
    return shift

def get_covarmatrix(x,xinterp,z,k,breakwidth):
    """
    Calculates the covariance matrix when needed.  Assumes both an
    interpolation and a smoothing----for now, only linear
    interpolation will work (only does one diagonal, but propagates
    the error correctly).
    """

    isort = sp.searchsorted(x,xinterp)
    f = (xinterp - x[isort -1 ])/(x[isort] - x[isort -1])
    z2 = sp.sqrt(
        (f**2)*(z[isort]**2) + ((1 - f)**2)*(z[isort - 1]**2)
        )
    #for a delta function
    if k.size == 1:
        return sp.diag(z2**2)

    covar1 = sp.zeros((z2.size ,z2.size ))
    covar2 = sp.zeros((z2.size ,z2.size ))

    #the part from interpolation---searchsorted has thrown out the
    #first index, the last index won't be selected because of the slice
    covar1[sp.diag_indices_from(covar1)] = z2**2
#    print sp.shape(f), sp.shape(z[isort.min():isort.max()-1]),sp.shape(z)
    diag1 = f[0:-1]*(1-f[0:-1])*(z[isort.min()  : isort.max() ]**2)

    #linear interpolation only has one diagonal
    covar1 += sp.diag(diag1,1)
    covar1 += sp.diag(diag1,-1)
    covar1 = sp.matrix(covar1)
                
    #the part from convolution
    #make kernel match size of input
    if k.size == z2.size  - 2:
        k = sp.r_[0,k,0]
    elif k.size == z2.size -1:
        k = sp.r_[0,k]
    #force it to wrap around
    cent = k.size//2
    k = sp.r_[ k[cent::], k[0:cent ] ]

    #for matrix equation, see Gardner 2003, Uncertainties in
    #Interpolated Spectral Data; equation 6

    # seems like there is a better way than a double loop....
    for i in range(z2.size):
        for n in range(int(breakwidth)):
            #can shorten the loop because most of the matrix is
            #zero:the idea is that kernels that are far apart
            #should have zero overlap.  If k2 gets shifted
            #relative to k1 more than 5x the kernel width, assume
            #no overlap
            j = i + n
            if j > z2.size - 1: j = z2.size -1
            k1 = sp.matrix(sp.roll(k,i))
            k2 = sp.matrix(sp.roll(k,j))
            trim1 = cent + i
            if trim1 <= k1.size:
                k1[:,trim1::] = 0
            else:
                k1[:,0: trim1 - k1.size] = 0

            trim2 = cent + j
            if trim2 <= k2.size:
                k2[:,trim2::] = 0
            else:
                k2[:,0: (trim2 - k2.size)] = 0

            covar2[i,j] = k1*covar1*k2.T
        #matrix is symmetric
    covar2 += covar2.T
    covar2[sp.diag_indices_from(covar2)] /= 2.
   
    return covar2


def metro_hast(ntrial,D,M,plot=False,keep=False):
    """
    This actualy does the work to fit the model to the data.  

    ntrial = number of iterations in the MCMC
    M = RescaleModel object (defined with the reference and smoothing kernel)
    D = Data (EmissionLine Object)

    plot=True will turn on interactive plotting---you can watch the
    chains as they progress!

    keep=True will return the Chain object used to store the MCMC,
    which can be saved latter (see do_map.py).
    

    """
    Mtry= deepcopy(M)
    chi2 = 1.e12
    chi2best = 1.e12

    pbest = deepcopy(M.p)
    accept = 0

    c = Chain()
    c.add(M,M(D))
    if plot ==1:
        plt.ion()
        c.plot()

    for i in range(ntrial):
        Mtry.p = M.step()
        chi2try = Mtry(D)
        
        if chi2try < chi2:
            
            M.p = deepcopy(Mtry.p)
            chi2 = deepcopy(chi2try)

            accept += 1
            c.add(M,chi2)
                
            if chi2 < chi2best:
                chi2best = deepcopy(chi2)
                pbest = deepcopy(M.p)

        else:
            prob = sp.exp(-chi2try/chi2)
            r = sp.rand()
            if r <= prob:
                M.p = deepcopy(Mtry.p)
                chi2 = deepcopy(chi2try)
                accept += 1
            c.add(M,chi2)
                
        if i%500 == 0 :
            print i,chi2best,chi2try
            if plot ==1:
                c.plot()

    if keep == 1:
        return chi2best,pbest,accept/float(ntrial),c
    else:
        return chi2best,pbest,accept/float(ntrial)


